Makefile.common

```diff
@@ -8,21 +8,62 @@ define newline
 
 endef
 
-HACL_HOME 	?= $(shell pwd)
+ifndef HACL_HOME
+HACL_HOME 	:= $(shell pwd)
+endif
 
 # Put your local configuration (e.g. HACL_HOME, KRML_HOME, etc.) in
 # Makefile.config
 -include $(HACL_HOME)/Makefile.config
 
-KRML_HOME 	?= $(HACL_HOME)/../karamel
-FSTAR_HOME	?= $(HACL_HOME)/../FStar
-VALE_HOME	?= $(HACL_HOME)/../vale
+ifndef KRML_HOME
+# assuming an Everest source tree
+KRML_HOME 	:= $(HACL_HOME)/../karamel
+endif
+
+ifndef FSTAR_HOME
+  FSTAR_EXE=$(shell which fstar.exe)
+  ifeq ($(FSTAR_EXE),)
+    # F* not found in PATH, assuming an Everest source tree
+    FSTAR_HOME := $(HACL_HOME)/../FStar
+  else
+    # F* found in PATH, assuming directory name ends with /bin
+    FSTAR_HOME := $(dir $(FSTAR_EXE))/..
+  endif
+endif
+
+ifndef FSTAR_ULIB
+FSTAR_ULIB := $(shell if test -d "$(FSTAR_HOME)/ulib" ; then echo "$(FSTAR_HOME)/ulib" ; else echo "$(FSTAR_HOME)/lib/fstar" ; fi)
+endif
+
+ifndef VALE_HOME
+# assuming an Everest source tree
+VALE_HOME	:= $(HACL_HOME)/../vale
+endif
```

Comment:
is the ifndef syntax better than `?=`?

Reply:
From https://www.gnu.org/software/make/manual/html_node/Conditional-Assignment.html: `FOO ?= bar` is equivalent to:
```
ifeq ($(origin FOO), undefined)
  FOO = bar
endif
```
which uses `=` instead of `:=`, thus preventing overwriting the variable when I sanitize it on lines 56 sqq.

Reply:
Great. Thanks for clarifying!